### PR TITLE
CConnection: Modify UserPasswdGetter and UserMsgBox interface to CConnection

### PR DESCRIPTION
**Pull request by KangLin**
CConnection: Modify UserPasswdGetter and UserMsgBox interface to CConnection
    
Problems with the original code: A process can only establish one connection.
After modification, multiple connections can be supported.

### CODE REVIEW

**Comment by CendioOssman** [comment]
Multiple connections is not something vncviewer can do presently, so I'm unclear for the need to refactor this?

**Comment by KangLin** [comment]
Multiple connections is not something vncviewer can do presently, so I'm unclear for the need to refactor this?

**Comment by KangLin** [comment]
But I need multiple connections. So refactored it.
It's good for degree architecture.

在 2024-07-30 15:33:31，"Pierre Ossman (ThinLinc team)" ***@***.***> 写道：


Multiple connections is not something vncviewer can do presently, so I'm unclear for the need to refactor this?

**Comment by CendioOssman** [comment]
Ah, that's right. You were using TigerVNC as a library.

I just want to reiterate that this is not something we support, so it's best if you can align your needs with those of the TigerVNC project.

That said, I see no need for these callbacks to be global, so this clean-up should be acceptable.

File: common/rfb/CConnection.h [hunk 1]
```diff
@@ -37,7 +37,13 @@ namespace rfb {
   class CSecurity;
   class IdentityVerifier;
 
-  class CConnection : public CMsgHandler {
+  class CConnection
+      : public CMsgHandler
+      , public UserPasswdGetter
+#if defined(HAVE_GNUTLS) || defined(HAVE_NETTLE)
+      ,public UserMsgBox
+#endif
```
Comment (CendioOssman) [inline common/rfb/CConnection.h:45]: These interfaces are overkill now that the callbacks are part of an existing class. So please put the methods directly on `CConnection` and get rid of `USerPasswdGetter` and `UserMsgBox`.

File: tests/perf/decperf.cxx [hunk 2]
```diff
@@ -85,6 +90,15 @@ class CConn : public rfb::CConnection {
...
 
+bool CConn::showMsgBox(int, const char *, const char *)
+{
+    return true;
+}
+
+void CConn::getUserPasswd(bool, std::string *, std::string *)
+{
+}
+
```
Comment (CendioOssman) [inline tests/perf/decperf.cxx:101]: Please try to maintain the ordering of methods in this file.

File: tests/perf/encperf.cxx [hunk 3]
```diff
@@ -118,8 +121,18 @@ class CConn : public rfb::CConnection {
...
 
+bool CConn::showMsgBox(int, const char *, const char *)
+{
+    return true;
+}
+
+void CConn::getUserPasswd(bool, std::string *, std::string *)
+{
+}
+
```
Comment (CendioOssman) [inline tests/perf/encperf.cxx:135]: Same thing here.

**Review by CendioOssman — CHANGES_REQUESTED** [review]

**Comment by KangLin** [comment]
I'm glad you can accept tigervnc as a library that supports multiple operating systems. I will help you achieve it.

File: common/rfb/CConnection.h [hunk 4]
```diff
@@ -36,8 +36,22 @@ namespace rfb {
...
+      M_ICONQUESTION = 0x20,
+      M_ICONWARNING = 0x30,
+      M_ICONINFORMATION = 0x40,
+      M_DEFBUTTON1 = 0,
+      M_DEFBUTTON2 = 0x100
+  };
+  
+  class CConnection
+      : public CMsgHandler
+  {
```
Comment (CendioOssman) [inline common/rfb/CConnection.h:54]: This line doesn't need to be changed anymore.

File: common/rfb/CConnection.h [hunk 5]
```diff
@@ -126,7 +140,15 @@ namespace rfb {
...
     // Methods to be overridden in a derived class
-
+    
+    // getUserPasswd gets the username and password.  This might involve a
+    // dialog, getpass(), etc.  The user buffer pointer can be null, in which
+    // case no user name will be retrieved.
+    virtual void getUserPasswd(bool secure, std::string* user,
+                               std::string* password);
+    
+    virtual bool showMsgBox(MsgBoxFlags flags, const char *title, const char *text);
```
Comment (CendioOssman) [inline common/rfb/CConnection.h:150]: Please keep these as pure methods.

**Review by CendioOssman — CHANGES_REQUESTED** [review]
Almost there. Just a couple small tweaks and I can merge this.

File: common/rfb/CConnection.h (hunk 4 above)
Comment (CendioOssman) [reply common/rfb/CConnection.h:53]: I'm afraid this line is still changed needlessly.

File: common/rfb/CSecurityRSAAES.cxx [hunk 6]
```diff
@@ -215,8 +214,8 @@ void CSecurityRSAAES::verifyServer()
     "Fingerprint: %02x-%02x-%02x-%02x-%02x-%02x-%02x-%02x\n"
     "Please verify that the information is correct and press \"Yes\". "
     "Otherwise press \"No\"", f[0], f[1], f[2], f[3], f[4], f[5], f[6], f[7]);
-  if (!msg->showMsgBox(UserMsgBox::M_YESNO, title, text.c_str()))
-    throw Exception("server key mismatch");
+  if (!cc->showMsgBox(MsgBoxFlags::M_YESNO, title, text.c_str()))
+    throw AuthFailureException("server key mismatch");
```
Comment (CendioOssman) [inline common/rfb/CSecurityRSAAES.cxx:218]: This last line should not have been changed.

File: common/rfb/CSecurityTLS.cxx [hunk 7]
```diff
@@ -472,8 +471,8 @@ void CSecurityTLS::checkSession()
                     "\n"
                     "Do you want to make an exception for this "
                     "server?", info.data);
-
-      if (!msg->showMsgBox(UserMsgBox::M_YESNO,
+
```
Comment (CendioOssman) [inline common/rfb/CSecurityTLS.cxx]: Some stray whitespace here.

File: vncviewer/UserDialog.h [hunk 8]
```diff
@@ -32,17 +30,10 @@ class UserDialog : public rfb::UserPasswdGetter,
...
 
   // UserMsgBox callbacks
-
-  bool showMsgBox(int flags, const char* title, const char* text) override;
-
-  void resetPassword();
-
- private:
-  std::string savedUsername;
-  std::string savedPassword;
```
Comment (CendioOssman) [inline vncviewer/UserDialog.h:45]: These changes should not be reverted.

File: common/rfb/CConnection.h [hunk 9]
```diff
@@ -145,7 +145,12 @@ namespace rfb {
     // case no user name will be retrieved.
     virtual void getUserPasswd(bool secure, std::string* user,
                                std::string* password) = 0;
-    
+    /*!
+     * Get x509 authentication CA and CRL file, the file format is pem.
+     * \param ca: certificate authority, the file format is pem.
+     * \param crl: certificate revocation list, the file format is pem.
+     */
+    virtual int getX509File(std::string* ca, std::string* crl) = 0;
```
Comment (CendioOssman) [inline common/rfb/CConnection.h:153]: This seems very unrelated to this PR?

**Review by CendioOssman — CHANGES_REQUESTED** [review]

File: common/rfb/CConnection.h (hunk 9 above)
Comment (KangLin) [reply common/rfb/CConnection.h:153]: Yes. Another similar issue, so I'll put it together.
 
CSecurityTLS: change the variable that x509 authentication CA and CRL file from global to local

File: common/rfb/CSecurityTLS.cxx (hunk 7 above)
Comment (CendioOssman) [reply common/rfb/CSecurityTLS.cxx]: This whitespace is still here.

File: common/rfb/CConnection.h [hunk 10]
```diff
@@ -35,6 +35,19 @@ namespace rfb {
   class CMsgReader;
   class CMsgWriter;
   class CSecurity;
+  class IdentityVerifier;
```
Comment (CendioOssman) [inline common/rfb/CConnection.h:38]: This line is also a revert of a previous commit.

File: vncviewer/UserDialog.h (hunk 8 above)
Comment (CendioOssman) [reply vncviewer/UserDialog.h:45]: This is still an issue.

File: common/rfb/CConnection.h (hunk 9 above)
Comment (CendioOssman) [reply common/rfb/CConnection.h:147]: I'm afraid I don't see them as similar. Please open a new PR with a rationale for why this change is needed.

**Review by CendioOssman — CHANGES_REQUESTED** [review]

File: vncviewer/CConn.h [hunk 11]
```diff
@@ -76,6 +82,8 @@ class CConn : public rfb::CConnection
   void handleClipboardAnnounce(bool available) override;
   void handleClipboardData(const char* data) override;
 
+  void reset_password_data();
```
Comment (CendioOssman) [inline vncviewer/CConn.h:85]: Please change this to `resetPassword()` as we use different naming style for methods and plain functions.

File: common/rfb/CConnection.h [hunk 12]
```diff
@@ -121,6 +133,21 @@ namespace rfb {
...
     // Methods to be overridden in a derived class
 
+    /*!
+     * \brief getUserPasswd gets the username and password.
+     *        This might involve a dialog, getpass(), etc.
+     * \param secure: Indicates whether this connection is secure
+     * \param user: user name buffer pointer. the pointer may be nullptr,
+     *              in which case no user name will be retrieved.
+     * \param password: password buffer pointer.
+     */
```
Comment (CendioOssman) [inline common/rfb/CConnection.h:143]: I'm afraid these documentation tags are not something we use. Please keep the existing style.

**Review by CendioOssman — CHANGES_REQUESTED** [review]
This is looking really good now. Nice work!

Just two small style changes and I can go ahead and merge this.

**Comment by CendioOssman** [comment]
Committed as 1b0387a36cf5f22a34b955d9c3f7ff0860e7467c. Thanks!

**Pull request closed without merging**